### DISCUSSION TITLE
`lua`/`vim9script`/`vim9class` issues/question from a `neovim` newbie

### DISCUSSION THREAD

**Discussion by errael** [neovim/neovim#30291] [General]
I'm still working to grok neovim and its ecosystem. I have some `vim9script` things I'd like to run under `nvim`. Looking at [vim9jit](https://github.com/tjdevries/vim9jit) to assist, I realized that almost everything, `lua`, `rust` and `neovim`, are all new to me; I never learned `vimscript`. (Had great fun opening a PR to `vim9jit`.)

1. `Neovim`'s `lua` is 5.1 acc'd to `nvim` doc; on `luajit`'s site I can't find a version of `lua` which `luajit` claims to be compatible with. It seems to be generally agreed that `luajit` is `5.1` compatible but... I'm considering getting a **book**. There is `Programming in Lua, Second Edition` which targets `lua 5.1`. For use with `neovim` is there any reason to get a newer version because of better/more of this/that, knowing that some things may not be applicable? Any other book recommended?
2. Installed `nvim-lspconfig` and `lua-language-server`. Saw the `---` typing information and how `LuaLS` uses that; cool. How can I generate a text file with all the errors, or just the type errors, from a project and/or a group of files?
3. For a lua file, trouble getting lsp's rename to work; but this things in that area seem to be in flux. Should this work? Did I screw up or is this expected in the future?
4. is there a way to run `neovim lua plugin` on `vim`? (found, in FAQ, that can't run `vim lua plugin` on `neovim`) What would it take to achieve this?
5. https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#lua_ls says to do something that doesn't work
   ```
   view this file in Nvim by running :help lspconfig-all
   ```
6. Consider (this is not `neovim` related, but I'm stymied)
    ```
    function G(...)
        local i = 2
        xxx = select(i, ...)                -- <<<<<<< WHY?
        print('    ', select(i, ...))
        print('    ', (select(i, ...)))     -- <<<<<<< WHY?
        print('    ', xxx)
    end
    G('a', 'b', 'c')
    ```
    ```
    	b	c
    	b
    	b
    ```
    I understand that `select(i, ...)` is the i-th arg to the last arg. why do the lines marked `WHY` work? I saw a reference to `(select(i, ...))` which I can no longer find. Is it magic that only has to do with varargs? Is there a general/deeper `lua` meaning?

My primary interest is to get a plugin, [splice9](https://github.com/errael/splice9), running on `nvim`. Ideally the same (or almost same) plugin source could run on both systems. (4) above is important (at least to me).

Two possibilities that make the most sense to me are
1. a `neovim` plugin that runs `vim9script`
2. a `vim` plugin that runs `neovim`'s flavor of `lua`

I like (2) the most. In conjunction with that it would be great to have a transpiler of `vim9script` to `lua`. I'm thinking it's only run once and the code is maintained as `lua` from then on. A key part is adding type information with `---` annotations. Some things won't/can't be translated (consider `vim9script`'s [null handling](https://vimhelp.org/vim9.txt.html#null-anomalies)). For unhandled or problematic constructs, I like putting markers in the target files that would generate compile error; another approach is to produce a report; and/or both. And a transpiler could have options for specific issues. I'm inclined to start with `antlr4` or `llvm`.

**Comment by clason** [comment]
That's a lot of questions all over the ecosystem (i.e., some of them are out of scope here), so you'll only get superficial answers to each:

1. Neovim supports Lua 5.1 _and Lua 5.1_ only; LuaJIT implements Lua 5.1 plus some extensions (which are _not_ supported).
2. `:h vim.diagnostic.get` and `vim.print()` the table (or use `:h vim.diagnostic.setqflist()` and work from there). Lua-language-server has a `--check` mode, see https://luals.github.io/wiki/usage/.
3. It should work, but not every server supports every feature in every situation
4. No. Not if they're written in Lua and/or make use of Neovim APIs.
5. Gotta open an issue there.
6. `:h select()`. Yes, this is a standard Lua function for extracting individual varargs.

Regarding your primary interest: There is vim9jit, which you have already discovered, and that is it. There is currently no real interest from the maintainers in adding vim9script support beyond that. Things _may_ change if (and only if) there appear some "must-have" plugins written in it, but I'm not holding my breath.

>t it would be great to have a transpiler of vim9script to lua. I'm thinking it's only run once and the code is maintained as lua from then on. 

That's exactly what vim9jit is and does.

**Comment by errael** [comment]
**About getting `vim9script` to run as/under `lua`/`neovim`**

The idea is to have a single source code that runs on both `neovim` and `vim`. I don't consider `vimscript` suitable. And `lua` with it's language support and maturity, is a great choice. `vim9script` has a long way to go to achieve parity with `lua` (the language is complete, there's refactoring support, it's relatively well known).

@tjdevries @clason I'm new to neovim... I can only guess at how much I'm not understanding or getting wrong.

> > 4. [...] run neovim lua plugin on vim? What would it take to achieve this?
> 4. No. Not if they're written in Lua and/or make use of Neovim APIs.

Yeah. It does seem that the crux of running the same lua code on both platforms is getting the `neovim` APIs working under `vim`. But what does that exactly mean?
- Does "Neovim APIs" mean the lua modules rooted at `vim.`?
- How much of this is `lua` versus `C`?
- How much of this depends on things new and/or unique to `neovim`?
- Would it be a lot of work to patch `vim` (if needed) for some `neovim` API support?
- Is there a lot of `neovim` API that could not be made readily available on `vim`?

> Regarding your primary interest: There is vim9jit ...
> 
> > it would be great to have a transpiler of vim9script to lua. I'm thinking it's only run once and the code is maintained as lua from then on.
> 
> That's exactly what vim9jit is and does.

`vim9jit` does not do that, AFAICT. And I probably should have said "translate" and not "transpile". As stated, what I'm looking for is

> the code is maintained as lua from then on

Doesn't `jit` imply that when running `neovim` you can do some some command in `neovim` like "run-vim9script" and it will read/compile/execute it on the fly? And if performance is an issue, consider that `luajit` currently produces code like
```
F = function(a, b, c)
  return vim9.ops.Plus(vim9.ops.Plus(a, b), c)
end
```
I may be misunderstanding the goals of `vim9jit`; its [README.md](https://github.com/tjdevries/vim9jit#vim9jit) says

> It will not be a full vimscript parser, but instead will only parse the necessary amount
> of vim9script to send the rest of the vimscript to Neovim's vimscript interpreter.

Consider there is class/interface to deal with and issues around integer/float handling and who knows what else. I don't think a fully automatic approach can work very well without a huge effort. How much of a vim9script runtime/VM has to be implemented for neovim?

Instead of a transpiler, an alternate approach is a `vim9script` to `lua` translation tool. Such a tool could
- create a `lua` file that uses the `lua` runtime/jit (not the vimscript runtime)
- add type information, `---@ annotations`, for use with development tools.
- mark problem areas that require manual fix up.
- add `if on_vim() ... endif` automatically where it's obviously needed
  Maybe an option to flag instead of automatically adding.

After the code is running on `neovim` the `vim9script` source can be put on the shelf.

To get the same `lua` file to run on both platforms, some of `neovim`'s lua API has to be made available on `vim`; just a subset I suspect. Hopefully most platform differences can be hidden in ported lua modules rooted at `vim.`. Maybe `if on_neovim() ... endif` is only needed in cases where the editing command itself has different semantics (if there are any) or to take advantage of `neovim` only features.

Instead of working to automatically run `vim9script`. This could well be a simpler approach, requiring less effort and with a better, more reliable outcome.

**Reply by clason** [reply]
>  I don't consider vimscript suitable.

But that is the only option on the table. (Legacy) Vimscript is the *lingua franca* of the ecosystem. If you want Lua, then write for Neovim. (Again, the issue is not the language but the different API and functionality -- look at the hoops people have to jump through to get popups/floating windows or jobs working on both, now add to that tree-sitter and lsp and libuv and...)

> I may be misunderstanding the goals of vim9jit

Apparently so; what I wrote is correct. `vim9jit` is an external tool that converts vim9script files into Lua files (with the help of a compatibility shim -- written in Lua to work around the former's type system -- the less said of which, the better). 

But it's goals are _not_ for writing code that runs on both platforms; the goal is _only_ to take Vim9 plugins (or, more accurately, rewritten runtime files) and convert them into files that Neovim can use.

There is currently _zero_ interest in extending that. You may want to, but then you're on your own.

There is even _less_ interest in adding a vim9script parser/engine to Neovim itself (which is what you seem to think vim9jit is? Don't get confused by the cute name; look at the code and documentation).

> Yeah. It does seem that the crux of running the same lua code on both platforms is getting the neovim APIs working under vim. But what does that exactly mean?

That is too much to answer here; please read the Neovim documentation (`:h api`, `:h lua`, `:h lua-guide`).

> Hopefully most platform differences can be hidden in ported lua modules rooted at vim..

Nope. There's way too much C-level code for that. What you're asking for is basically for Vim to bundle Neovim -- good luck with that.

**Comment by errael** [comment]
> > 1. ...I'm considering getting a book. There is _Programming in Lua, Second Edition_
> >   which targets lua 5.1. For use with neovim is there any reason to get a newer version
> >   because of better/more of this/that, knowing that some things may not be applicable?
> >   Any other book recommended? 

Any ideas?

> 2. `:h vim.diagnostic.get` and `vim.print()` the table (or use `:h vim.diagnostic.setqflist()` and work from there). Lua-language-server has a `--check` mode, see https://luals.github.io/wiki/usage/.

Thanks, works good.

> > 3. trouble getting lsp's rename to work
> 3. It should work, but not every server supports every feature in every situation

Oops. I was using v0.10.1, it does work in v0.11.0-dev. Refactoring, very cool.

> 6. `:h select()`. Yes, this is a standard Lua function for extracting individual varargs.

My confusion/misunderstanding comes from the "extra" `()` in the following
```
print((select(1, ...)))
```
Is that construct specific to select?

**Reply by telemachus** [reply]
> My confusion/misunderstanding comes from the "extra" `()` in the following
> 
> ```
> print((select(1, ...)))
> ```
> 
> Is that construct specific to select?

No: extra parentheses limit the number of return values to one. From [Lua's documentation](https://www.lua.org/manual/5.1/manual.html#2.5):

> Any expression enclosed in parentheses always results in only one value. Thus, `(f(x,y,z))` is always a single value, even if `f` returns several values. (The value of `(f(x,y,z))` is the first value returned by `f` or `nil` if `f` does not return any values.)

Re books, *Programming in Lua* is outstanding. (Yes, I would go with the second edition if you care primarily or only about neovim.) [*Beginning Lua Programming*](https://www.amazon.com/Beginning-Lua-Programming-Kurt-Jung/dp/0470069171) is fine (and if I remember right, it was also written for 5.1), but much more elementary. If you already know how to program, go with *Programming in Lua*. There is [a print copy of the reference manual for Lua 5.1](https://www.amazon.com/Lua-Reference-Manual-Roberto-Ierusalimschy/dp/8590379833), but you can read [that for free online](https://www.lua.org/manual/5.1).

**Reply by clason** [reply]
or `:h luaref` (yes, make sure to stick to Lua 5.1 -- later "versions" are different language dialects and will not work in Neovim).

**Reply by errael** [reply]
> No: extra parentheses limit the number of return values to one

Aha, missed that. An `expr`.  Now I see why the following is an error
```
zzz = (7, 5, 6)
```
And more thanks for the book feedback, will order.

**Comment by errael** [comment]
> _Programming in Lua_ is outstanding ... the second edition

This book recommendation from @telemachus  is the winner in this thread.

vim9jit may have been useful for translating simplest vim9script from vim9.0. (it's pretty buggy)
